usr/lib/common/mech_sha.c

```diff
@@ -524,3 +525,246 @@ CK_RV ckm_generic_secret_key_gen(STDLL_TokData_t *tokdata, TEMPLATE *tmpl)
+            break;
+        default:
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCONSISTENT));
+            return CKR_TEMPLATE_INCONSISTENT;
+        }
+        break;
+    default:
+        TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCONSISTENT));
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+
+    if (derived_keylen == 0)
+        derived_keylen = allowed_keysize;
+
+    if (derived_keylen > hsize || derived_keylen > allowed_keysize) {
+        TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCONSISTENT));
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+
+    if (!template_get_class(base_key_obj->template, &base_key_class,
+                            &base_key_type)) {
+        TRACE_ERROR("Could not find CKA_CLASS in the template\n");
+        return CKR_TEMPLATE_INCOMPLETE;
+    }
+
+    if (base_key_class != CKO_SECRET_KEY) {
+        TRACE_ERROR("Base key is not a secret key\n");
+        return CKR_KEY_TYPE_INCONSISTENT;
+    }
+
+    switch (base_key_type) {
+    case CKK_GENERIC_SECRET:
+    case CKK_DES:
+    case CKK_DES2:
+    case CKK_DES3:
+    case CKK_AES:
+    case CKK_AES_XTS:
+        break;
+    default:
+        TRACE_ERROR("Base key type is not supported\n");
+        return CKR_KEY_TYPE_INCONSISTENT;
+    }
+
+    /* Digest the base key to derive the derived key */
+    rc = template_attribute_get_non_empty(base_key_obj->template,
+                                          CKA_VALUE, &base_key_value);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_VALUE for the base key.\n");
+        return rc;
+    }
+
+    rc = digest_mgr_init(tokdata, sess, &ctx, &digest_mech, FALSE);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return rc;
+    }
+
+    rc = digest_mgr_digest(tokdata, sess, FALSE, &ctx,
+                           base_key_value->pValue, base_key_value->ulValueLen,
+                           derived_key_value, &hsize);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("digest_mgr_digest failed with rc = %s\n", ock_err(rc));
+        digest_mgr_cleanup(tokdata, sess, &ctx);
+        return rc;
+    }
+
+    rc = build_attribute(CKA_VALUE, derived_key_value, derived_keylen,
+                         &value_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to build the attribute from CKA_VALUE, rc=%s.\n",
+                    ock_err(rc));
+        return rc;
+    }
+
+    switch (derived_keytype) {
+    case CKK_GENERIC_SECRET:
+    case CKK_AES:
+    case CKK_AES_XTS:
+        /* Supply CKA_VAUE_LEN since this is required for those key types */
+        rc = build_attribute(CKA_VALUE_LEN, (CK_BYTE*)&derived_keylen,
+                             sizeof(derived_keylen), &vallen_attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Failed to build the attribute from CKA_VALUE_LEN, "
+                        "rc=%s.\n", ock_err(rc));
+            goto end;
+        }
+        break;
+    case CKK_DES:
```

Comment:
Are there no weak keys possible for DES2 and DES3? A DES2 key is just the concatenation of two DES keys, so we could check the two keys individually for weakness .... similar to DES3.